### ISSUE TITLE
Date format does not respect locale

### ISSUE THREAD

**Issue by 2-4601** [WhisperSystems/Signal-Android#3102]
Currently TS seems to display all dates in the format [MMM d, yyyy](https://github.com/WhisperSystems/TextSecure/blob/d8521637bb2a545d02d2b7a32127a4f01b6d0317/src/org/thoughtcrime/securesms/util/DateUtils.java) irrespective of the locale. The dates are translated correctly though.

Here's some information about date formats in various regions:
https://en.wikipedia.org/wiki/Date_format_by_country

TS 2.13.0
This issue does not depend on whether you have changed the system language or TS language. Neither work.

**Comment by agrajaghh** [comment]
If it's the same issue like #2684 it has been fixed in https://github.com/WhisperSystems/TextSecure/commit/d8521637bb2a545d02d2b7a32127a4f01b6d0317

**Comment by 2-4601** [comment]
That should be fixed in 2.13.0 beta release, right?

I searched only open issues before creating this so I didn't see #2684. But that issue seems to be about date translations and not about date format? Before 2.13.0 I also experienced that issue but it seems to be fixed now.

This issue is about date formatting which does not respect the locale. This was introduced in 2.13.0. I did not see it before. Possibly because #2684 prevented me from seeing translated dates.

For example the Finnish date format should be `d. MMM yyyy` and not `MMM d, yyyy`.

TextSecure shows it wrong. Android gets the date formatting right.
![svejk](https://cloud.githubusercontent.com/assets/174176/7457751/9f85e804-f297-11e4-9aff-4145e150e340.jpg)

**Comment by agrajaghh** [comment]
oh sry, yes the commit I mentioned is already in 2.13...

**Comment by rhodey** [comment]
ahhh bummer, I thought [this line](https://github.com/WhisperSystems/TextSecure/blob/master/src/org/thoughtcrime/securesms/util/DateUtils.java#L44) was supposed to take care of that. thanks for the expected vs actual test vectors, I'll look into it.

**Comment by rhodey** [comment]
```java
private static void testLocalizedSimpleDateFormat(String template, Locale locale) {
  Log.d(DateUtils.class.getName(), "LOCALE >> " + locale + ", TEMPLATE >> " + template);
  SimpleDateFormat dateFormat = new SimpleDateFormat(template, locale);
  Log.d(DateUtils.class.getName(), "LOCALE >> " + locale + ", LOCALE TEMPLATE >> " + dateFormat.toLocalizedPattern());
}
```

```
DateUtils﹕ LOCALE >> fi, TEMPLATE >> MMM d, yyyy HH:mm:ss zzz
DateUtils﹕ LOCALE >> fi, LOCALE TEMPLATE >> MMM d, yyyy HH:mm:ss zzz
```

:confused: Java and Android's date processing stuff is junk. We switched to `SimpleDateFormat` because `android.text.format.DateUtils` effectively ignores every `Locale` arg it is passed and now this. All internet advice tells us to use the new Java 8 APIs or `DateFormat.getDateInstance(int,Locale)`, where `int` refers to a predefined date format string ie no custom date formats.

**Comment by 2-4601** [comment]
I've banged my head on the wall and lost a lot of sleep when it comes to handling dates and timezones in code. Luckily I haven't had to do it with Java :)

The _mm/dd_ format doesn't really bother me that much. Especially when the month is displayed as a word (or abbreviation). If it were all numerals e.g. `05/04` it would be much more confusing for the _dd/mm_ part of the world. But the current format `May 4` (or the equivalent translation) is OK.

However the Finnish date in TS is pretty confusing because it uses the full month name in an inflected form that is only valid grammar when the format is _dd/mm_. So `toukokuuta 4` is not just wrong format but also improper language :)
The abbreviated format `Tou 4` would be OK. Even though it's not correct format it's still understandable (less invalid grammar). I don't know why the English month name gets abbreviated to the three first letters while the Finnish month is in the full inflected form. The template is supposed to be the same for both?

And this is just one language. Other languages might be broken too.

**Comment by unrulygnu** [comment]
Generally in Java, if the canned date formats (`SHORT`, `MEDIUM`, `LONG`, `FULL`) are insufficient and you want to use a custom date format, localization is painful. Element order, punctuation, abbreviations, prefixes/suffixes, etc. all vary wildly between languages and their respective date forms.  Developers are usually left to one of three options:

1. Create a separate format string for each locale (shoot me now)
2. Use a library like those in [ICU](http://site.icu-project.org/)
3. Settle for ugly localization and hope for the best (Signal's current state)

Android 4.3 (API 18, `JELLY_BEAN_MR2`) [added a method](https://developer.android.com/reference/android/text/format/DateFormat.html#getBestDateTimePattern%28java.util.Locale,%20java.lang.String%29) that makes option 2 native, implementing [ICU4C](https://ssl.icu-project.org/apiref/icu4c/) in the platform via JNI. I'm submitting a PR that implements this method.

The following images include cropped screenshots (Nexus 5, Android 6.0 MRA58K) in English, Korean, German, and Finnish, showing dates in the ConversationList, Conversation, and MessageDetails. @agrajaghh , @2-4601 , what do you think?

**Before**
![before](https://cloud.githubusercontent.com/assets/11031903/11263835/e24a886c-8ed1-11e5-83e3-310dde46ac38.png)

**After** (only Android 4.3 and above)
![after](https://cloud.githubusercontent.com/assets/11031903/11263837/e662d486-8ed1-11e5-9fbe-66726f7e6676.png)

**Comment by agrajaghh** [comment]
definitely better in german, it was just wrong before :blush:

**Comment by 2-4601** [comment]
Yaaasss! This is glorious work. This makes it correct. Thanks.

Any idea why the Finnish date is displayed with full month name (in an inflected form) while other languages display the short abbreviated form?

**Comment by unrulygnu** [comment]
@2-4601 
> Any idea why the Finnish date is displayed with full month name (in an inflected form) 

`MMM` is supposed to render an abbreviated textual month name in all languages. I suspect a dangerous combination of Finnish mischief and beer was responsible for this one. :stuck_out_tongue_closed_eyes: 

> Or is it more complicated

Unfortunately, it is.  One of the date formats, for instance, is assembled in steps using multiple conditionals.

**Comment by 2-4601** [comment]
solved in 3.6.0 for >= 4.3 devices
